File: app/src/main/cpp/engine/TransportClockAnchor.cpp
#include "TransportClockAnchor.h"

#include <algorithm>
#include <time.h>

namespace transport_clock {

int64_t monotonicNowNs() {
    timespec ts{};
    if (clock_gettime(CLOCK_MONOTONIC, &ts) != 0) {
        return 0;
    }
    return static_cast<int64_t>(ts.tv_sec) * 1'000'000'000LL +
           static_cast<int64_t>(ts.tv_nsec);
}

int64_t TransportClockAnchor::transportFrameAt(const int64_t monotonicNs) const {
    if (sampleRateHz <= 0 || monotonicStartNs <= 0) {
        return transportStartFrame;
    }
    const int64_t elapsedNs = monotonicNs - monotonicStartNs;
    const int64_t frameDelta =
        (elapsedNs * static_cast<int64_t>(sampleRateHz)) / 1'000'000'000LL;
    return transportStartFrame + frameDelta;
}

int64_t TransportClockAnchor::transportMsAt(const int64_t monotonicNs) const {
    if (sampleRateHz <= 0) {
        return 0;
    }
    const int64_t frame = transportFrameAt(monotonicNs) - transportStartFrame;
    return (std::max<int64_t>(0, frame) * 1000LL) / static_cast<int64_t>(sampleRateHz);
}

int64_t TransportClockAnchor::monotonicNsForTransportFrame(const int64_t frame) const {
    if (sampleRateHz <= 0 || monotonicStartNs <= 0) {
        return monotonicStartNs;
    }
    const int64_t frameDelta = frame - transportStartFrame;
    const int64_t nsDelta =
        (frameDelta * 1'000'000'000LL) / static_cast<int64_t>(sampleRateHz);
    return monotonicStartNs + nsDelta;
}

int64_t TransportClockAnchor::monotonicNsForTransportMs(
    const int64_t transportMsFromStartFrame) const {
    if (sampleRateHz <= 0) {
        return monotonicStartNs;
    }
    const int64_t frameDelta =
        (std::max<int64_t>(0, transportMsFromStartFrame) *
         static_cast<int64_t>(sampleRateHz)) /
        1000LL;
    return monotonicNsForTransportFrame(transportStartFrame + frameDelta);
}

} // namespace transport_clock

File: app/src/main/java/com/georgv/audioworkstation/core/audio/latency/TransportClockAnchor.kt
package com.georgv.audioworkstation.core.audio.latency

/** JVM mirror of native [transport_clock::TransportClockAnchor] for mapping tests and render-ahead math. */
data class TransportClockAnchor(
    val transportStartFrame: Long,
    val monotonicStartNs: Long,
    val sampleRateHz: Int,
) {
    val isValid: Boolean
        get() = sampleRateHz > 0 && monotonicStartNs > 0L

    fun transportFrameAt(monotonicNs: Long): Long {
        if (!isValid) return transportStartFrame
        val elapsedNs = monotonicNs - monotonicStartNs
        val frameDelta = elapsedNs * sampleRateHz / NsPerSecond
        return transportStartFrame + frameDelta
    }

    fun transportMsAt(monotonicNs: Long): Long {
        if (sampleRateHz <= 0) return 0L
        val frame = transportFrameAt(monotonicNs) - transportStartFrame
        return frame.coerceAtLeast(0L) * 1000L / sampleRateHz
    }

    fun monotonicNsForTransportFrame(frame: Long): Long {
        if (!isValid) return monotonicStartNs
        val frameDelta = frame - transportStartFrame
        val nsDelta = frameDelta * NsPerSecond / sampleRateHz
        return monotonicStartNs + nsDelta
    }

    fun monotonicNsForTransportMs(transportMsFromStartFrame: Long): Long {
        if (sampleRateHz <= 0) return monotonicStartNs
        val frameDelta = transportMsFromStartFrame.coerceAtLeast(0L) * sampleRateHz / 1000L
        return monotonicNsForTransportFrame(transportStartFrame + frameDelta)
    }

    private companion object {
        const val NsPerSecond = LatencyTimeConstants.NanosecondsPerSecond
    }
}
